"No skolem info" error when via type mentions out-of-scope type variable

```
λ> newtype Foo a = Foo a deriving Show via (Identity b)

<interactive>:10:32: error:ghc-stage2: panic! (the 'impossible' happened)
 (GHC version 8.3.20170913 for x86_64-unknown-linux):
       No skolem info:
 b_a3wa
 Call stack:
     CallStack (from HasCallStack):
       callStackDoc, called at compiler/utils/Outputable.hs:1142:37 in ghc:Outputable
       pprPanic, called at compiler/typecheck/TcErrors.hs:2844:5 in ghc:TcErrors
```

I think this will likely be fixed when I properly propagate the `via` type's type variables to the instance head itself.

I gave this a go in https://github.com/RyanGlScott/ghc/commit/217e491341bc4f991e9c890d8ad5eeb6877d988a. Alas, this causes some existing programs to break. As one example, if you try running the `deriving-via-standalone` program with that commit, you get:

```
deriving-via-standalone.hs:17:1: error:
    The exact Name ‘k1_a3K8’ is not in scope
      Probable cause: you used a unique Template Haskell name (NameU), 
      perhaps via newName, but did not bind it
      If that's it, then -ddump-splices might be useful

deriving-via-standalone.hs:17:1: error:
    The exact Name ‘k1_a3K8’ is not in scope
      Probable cause: you used a unique Template Haskell name (NameU), 
      perhaps via newName, but did not bind it
      If that's it, then -ddump-splices might be useful

deriving-via-standalone.hs:17:1: error:
    The exact Name ‘k1_a3K8’ is not in scope
      Probable cause: you used a unique Template Haskell name (NameU), 
      perhaps via newName, but did not bind it
      If that's it, then -ddump-splices might be useful

deriving-via-standalone.hs:17:1: error:
    The exact Name ‘k1_a3K8’ is not in scope
      Probable cause: you used a unique Template Haskell name (NameU), 
      perhaps via newName, but did not bind it
      If that's it, then -ddump-splices might be useful

deriving-via-standalone.hs:17:1: error:
    The exact Name ‘k1_a3K8’ is not in scope
      Probable cause: you used a unique Template Haskell name (NameU), 
      perhaps via newName, but did not bind it
      If that's it, then -ddump-splices might be useful

deriving-via-standalone.hs:17:1: error:
    The exact Name ‘k1_a3K8’ is not in scope
      Probable cause: you used a unique Template Haskell name (NameU), 
      perhaps via newName, but did not bind it
      If that's it, then -ddump-splices might be useful

deriving-via-standalone.hs:17:1: error:
    The exact Name ‘k1_a3K8’ is not in scope
      Probable cause: you used a unique Template Haskell name (NameU), 
      perhaps via newName, but did not bind it
      If that's it, then -ddump-splices might be useful
```

I'm not sure yet what is causing this.

--- a/compiler/rename/RnSource.hs
+++ b/compiler/rename/RnSource.hs
@@ -25,11 +25,10 @@ import RdrName
 import RnTypes
 import RnBinds
 import RnEnv
-import RnUtils          ( HsDocContext(..), mapFvRn, mapMaybeFvRn
-                        , bindLocalNames, checkDupRdrNames, inHsDocContext
-                        , bindLocalNamesFV, checkShadowedRdrNames
-                        , warnUnusedTypePatterns, extendTyVarEnvFVRn
-                        , newLocalBndrsRn )
+import RnUtils          ( HsDocContext(..), mapFvRn, bindLocalNames
+                        , checkDupRdrNames, inHsDocContext, bindLocalNamesFV
+                        , checkShadowedRdrNames, warnUnusedTypePatterns
+                        , extendTyVarEnvFVRn, newLocalBndrsRn )
 import RnUnbound        ( mkUnboundName )
 import RnNames
 import RnHsDoc          ( rnHsDoc, rnMbLHsDoc )
@@ -939,13 +938,14 @@ Here 'k' is in scope in the kind signature, just like 'x'.
 -}
 
 rnSrcDerivDecl :: DerivDecl GhcPs -> RnM (DerivDecl GhcRn, FreeVars)
-rnSrcDerivDecl (DerivDecl ty deriv_strat overlap)
+rnSrcDerivDecl (DerivDecl ty mds overlap)
   = do { standalone_deriv_ok <- xoptM LangExt.StandaloneDeriving
        ; unless standalone_deriv_ok (addErr standaloneDerivErr)
-       ; (ty', deriv_strat', fvs)
-           <- rnLHsInstTypeAndThen ctx_text ty $
-              mapMaybeFvRn (rnLDerivStrategy (GenericCtx ctx_text)) deriv_strat
-       ; return (DerivDecl ty' deriv_strat' overlap, fvs) }
+       ; (mds', ty', fvs)
+           <- rnLDerivStrategy (GenericCtx ctx_text) mds $ \strat_tvs via_ty ->
+              rnAndReportFloatingViaTvs strat_tvs via_ty $
+              rnLHsInstType ctx_text ty
+       ; return (DerivDecl ty' mds' overlap, fvs) }
   where
     ctx_text = text "a deriving declaration"
 
@@ -1659,30 +1659,73 @@ rnLHsDerivingClause :: HsDocContext -> LHsDerivingClause GhcPs
 rnLHsDerivingClause doc
                 (L loc (HsDerivingClause { deriv_clause_strategy = dcs
                                          , deriv_clause_tys = L loc' dct }))
-  = do { (dcs', fvs1) <- mapMaybeFvRn (rnLDerivStrategy doc) dcs
-       ; (dct', fvs2) <- mapFvRn (rnHsSigType doc) dct
-       ; return ( L loc (HsDerivingClause { deriv_clause_strategy = dcs'
-                                          , deriv_clause_tys = L loc' dct' })
-                , fvs1 `plusFV` fvs2 ) }
+  = do { (dcs', dct', fvs)
+           <- rnLDerivStrategy doc dcs $ \strat_tvs via_ty ->
+              mapFvRn (rnAndReportFloatingViaTvs strat_tvs via_ty .
+                       rnHsSigType doc) dct
+       ; pure ( L loc (HsDerivingClause { deriv_clause_strategy = dcs'
+                                       , deriv_clause_tys = L loc' dct' })
+              , fvs ) }
 
-rnLDerivStrategy :: HsDocContext -> LDerivStrategy GhcPs
-                 -> RnM (LDerivStrategy GhcRn, FreeVars)
-rnLDerivStrategy doc (L loc ds)
-  = do { let extNeeded :: LangExt.Extension
-             extNeeded 
-               | ViaStrategy{} <- ds
-               = LangExt.DerivingVia
-               | otherwise
-               = LangExt.DerivingStrategies
+rnLDerivStrategy :: forall a.
+                    HsDocContext -> Maybe (LDerivStrategy GhcPs)
+                 -> ([Name] -> SDoc -> RnM (a, FreeVars))
+                 -> RnM (Maybe (LDerivStrategy GhcRn), a, FreeVars)
+rnLDerivStrategy doc mds thing_inside
+  = case mds of
+      Nothing -> boring_case Nothing
+      Just ds -> do (ds', thing, fvs) <- rn_deriv_strat ds
+                    pure (Just ds', thing, fvs)
+  where
+    rn_deriv_strat :: LDerivStrategy GhcPs
+                   -> RnM (LDerivStrategy GhcRn, a, FreeVars)
+    rn_deriv_strat (L loc ds) = do
+      let extNeeded :: LangExt.Extension
+          extNeeded
+            | ViaStrategy{} <- ds
+            = LangExt.DerivingVia
+            | otherwise
+            = LangExt.DerivingStrategies
 
-       ; unlessXOptM extNeeded $
-           failWith $ illegalDerivStrategyErr ds
-       ; case ds of
-           StockStrategy    -> pure (L loc StockStrategy,    emptyFVs)
-           AnyclassStrategy -> pure (L loc AnyclassStrategy, emptyFVs)
-           NewtypeStrategy  -> pure (L loc NewtypeStrategy,  emptyFVs)
-           ViaStrategy ty   -> do { (ty', fvs) <- rnHsSigType doc ty
-                                  ; pure (L loc (ViaStrategy ty'), fvs) } }
+      unlessXOptM extNeeded $
+        failWith $ illegalDerivStrategyErr ds
+
+      case ds of
+        StockStrategy    -> boring_case (L loc StockStrategy)
+        AnyclassStrategy -> boring_case (L loc AnyclassStrategy)
+        NewtypeStrategy  -> boring_case (L loc NewtypeStrategy)
+        ViaStrategy ty   ->
+          do (ty', thing, fvs) <- rnHsSigTypeAndThen doc ty $ \strat_tvs ->
+                                  thing_inside strat_tvs (ppr ty)
+             pure (L loc (ViaStrategy ty'), thing, fvs)
+
+    boring_case :: mds
+                -> RnM (mds, a, FreeVars)
+    boring_case mds = do
+      (thing, fvs) <- thing_inside [] empty
+      pure (mds, thing, fvs)
+
+-- TODO RGS: What is going on here?
+rnAndReportFloatingViaTvs
+  :: [Name] -> SDoc
+     -- TODO RGS: What are these first two arguments? Also, mention that
+     -- they're only used for error message purposes.
+  -> TcM (LHsSigType GhcRn, FreeVars)
+  -> RnM (LHsSigType GhcRn, FreeVars)
+rnAndReportFloatingViaTvs tv_names via_ty thing_inside
+  = do (thing@HsIB {hsib_body = L l _}, thing_fvs) <- thing_inside
+       setSrcSpan l $ mapM_ (report_floating_via_tv thing thing_fvs) tv_names
+       pure (thing, thing_fvs)
+  where
+    report_floating_via_tv :: LHsSigType GhcRn -> FreeVars -> Name -> RnM ()
+    report_floating_via_tv deriv_ty used_names tv_name
+      = unless (tv_name `elemNameSet` used_names) $ addErr $ vcat
+          [ text "Type variable" <+> quotes (ppr tv_name) <+>
+            text "is bound in the" <+> quotes (text "via") <+>
+            text "type" <+> quotes via_ty
+            -- TODO RGS: Should we say _where_ it needs to be mentioned?
+          , text "but is not mentioned in" <+> quotes (ppr deriv_ty) <>
+            text ", which is illegal" ]
 
 badGadtStupidTheta :: HsDocContext -> SDoc
 badGadtStupidTheta _
@@ -1694,9 +1737,9 @@ illegalDerivStrategyErr ds
   = vcat [ text "Illegal deriving strategy" <> colon <+> derivStrategyName ds
          , text enableStrategy ]
 
-  where 
+  where
     enableStrategy :: String
-    enableStrategy 
+    enableStrategy
       | ViaStrategy{} <- ds
       = "Use DerivingVia to enable this extension"
       | otherwise

--- a/compiler/rename/RnTypes.hs
+++ b/compiler/rename/RnTypes.hs
@@ -13,7 +13,7 @@ module RnTypes (
         rnHsKind, rnLHsKind,
         rnHsSigType, rnHsSigTypeAndThen, rnHsWcType,
         rnHsSigWcType, rnHsSigWcTypeScoped,
-        rnLHsInstType, rnLHsInstTypeAndThen,
+        rnLHsInstType,
         newTyVarNameRn, collectAnonWildCards,
         rnConDeclFields,
         rnLTyVar,
@@ -271,11 +271,12 @@ rnHsSigType :: HsDocContext -> LHsSigType GhcPs
 -- that cannot have wildcards
 rnHsSigType ctx hs_sig_ty
   = do { (hs_sig_ty', _, fvs) <- rnHsSigTypeAndThen ctx hs_sig_ty
-                                                    (pure ((), emptyFVs))
+                                                    (\_ -> pure ((), emptyFVs))
        ; pure (hs_sig_ty', fvs) }
 
 rnHsSigTypeAndThen
-  :: HsDocContext -> LHsSigType GhcPs -> RnM (a, FreeVars)
+  :: HsDocContext -> LHsSigType GhcPs
+  -> ([Name] -> RnM (a, FreeVars))
   -> RnM (LHsSigType GhcRn, a, FreeVars)
 rnHsSigTypeAndThen ctx (HsIB { hsib_body = hs_ty }) thing_inside
   = do { traceRn "rnHsSigType" (ppr hs_ty)
@@ -283,7 +284,7 @@ rnHsSigTypeAndThen ctx (HsIB { hsib_body = hs_ty }) thing_inside
        ; ((hs_sig_ty, thing), fvs)
            <- rnImplicitBndrs (not (isLHsForAllTy hs_ty)) ctx vars $ \ vars ->
               do { (body', fvs1) <- rnLHsType ctx hs_ty
-                 ; (thing, fvs2) <- thing_inside
+                 ; (thing, fvs2) <- thing_inside vars
                  ; pure ( (mk_implicit_bndrs vars body' fvs1, thing)
                         , fvs1 `plusFV` fvs2 ) }
        ; pure (hs_sig_ty, thing, fvs) }
@@ -339,27 +340,19 @@ rnLHsInstType :: SDoc -> LHsSigType GhcPs -> RnM (LHsSigType GhcRn, FreeVars)
 -- Rename the type in an instance or standalone deriving decl
 -- The 'doc_str' is "an instance declaration" or "a VECTORISE pragma"
 rnLHsInstType doc_str inst_ty
-  = do { (inst_ty', _, fvs) <- rnLHsInstTypeAndThen doc_str inst_ty
-                                                    (pure ((), emptyFVs))
-       ; pure (inst_ty', fvs) }
-
-rnLHsInstTypeAndThen
-  :: SDoc -> LHsSigType GhcPs -> RnM (a, FreeVars)
-  -> RnM (LHsSigType GhcRn, a, FreeVars)
-rnLHsInstTypeAndThen doc_str inst_ty thing_inside
   | Just cls <- getLHsInstDeclClass_maybe inst_ty
   , isTcOcc (rdrNameOcc (unLoc cls))
          -- The guards check that the instance type looks like
          --   blah => C ty1 .. tyn
   = do { let full_doc = doc_str <+> text "for" <+> quotes (ppr cls)
-       ; rnHsSigTypeAndThen (GenericCtx full_doc) inst_ty thing_inside }
+       ; rnHsSigType (GenericCtx full_doc) inst_ty }
 
   | otherwise  -- The instance is malformed, but we'd still like
                -- to make progress rather than failing outright, so
                -- we report more errors.  So we rename it anyway.
   = do { addErrAt (getLoc (hsSigType inst_ty)) $
          text "Malformed instance:" <+> ppr inst_ty
-       ; rnHsSigTypeAndThen (GenericCtx doc_str) inst_ty thing_inside }
+       ; rnHsSigType (GenericCtx doc_str) inst_ty }
 
 mk_implicit_bndrs :: [Name]  -- implicitly bound
                   -> a           -- payload

--- a/compiler/typecheck/TcDeriv.hs
+++ b/compiler/typecheck/TcDeriv.hs
@@ -612,27 +612,23 @@ deriveStandalone (L loc (DerivDecl deriv_ty mbl_deriv_strat overlap_mode))
     addErrCtxt (standaloneCtxt deriv_ty)  $
     do { traceTc "Standalone deriving decl for" (ppr deriv_ty)
        ; let mb_deriv_strat = fmap unLoc mbl_deriv_strat
-       ; traceTc "Standalone deriving (pre-typechecking)" $
-           vcat [ text "mb_deriv_strat" <+> ppr mb_deriv_strat
-                , text "deriv_ty"       <+> ppr deriv_ty ]
-       ; (cls_tvs, theta, cls, inst_tys) <-
-           tcHsClsInstType TcType.InstDeclCtxt deriv_ty
-       ; (strat_tvs, mb_deriv_strat') <-
-           tcExtendTyVarEnv cls_tvs $
-           maybe ([], Nothing) (fmap Just) <$>
-           traverse tcDerivStrategy mb_deriv_strat
-       ; traceTc "Standalone deriving (post-typechecking)" $ vcat
-              [ text "cls_tvs:" <+> ppr cls_tvs
+       ; traceTc "Deriving strategy (standalone deriving)" $
+           vcat [ppr mb_deriv_strat, ppr deriv_ty]
+       ; (mb_deriv_strat', tvs, (theta, cls, inst_tys))
+           <- tcDerivStrategy mb_deriv_strat $ do
+                (tvs, theta, cls, inst_tys)
+                  <- tcHsClsInstType TcType.InstDeclCtxt deriv_ty
+                pure (tvs, (theta, cls, inst_tys))
+       ; traceTc "Standalone deriving;" $ vcat
+              [ text "tvs:" <+> ppr tvs
               , text "theta:" <+> ppr theta
               , text "cls:" <+> ppr cls
               , text "tys:" <+> ppr inst_tys
-              , text "strat_tvs:" <+> ppr strat_tvs
               , text "mb_deriv_strat:" <+> ppr mb_deriv_strat' ]
                 -- C.f. TcInstDcls.tcLocalInstDecl1
        ; checkTc (not (null inst_tys)) derivingNullaryErr
 
-       ; let tvs = cls_tvs `chkAppend` strat_tvs
-             cls_tys = take (length inst_tys - 1) inst_tys
+       ; let cls_tys = take (length inst_tys - 1) inst_tys
              inst_ty = last inst_tys
        ; traceTc "Standalone deriving:" $ vcat
               [ text "class:" <+> ppr cls
@@ -679,8 +675,8 @@ deriveTyData :: [TyVar] -> TyCon -> [Type]   -- LHS of data or data instance
 -- a no-op nowadays.
 deriveTyData tvs tc tc_args mb_deriv_strat deriv_pred
   = setSrcSpan (getLoc (hsSigType deriv_pred)) $  -- Use loc of the 'deriving' item
-    do  { (deriv_tvs, cls, cls_tys, cls_arg_kinds, _strat_tvs, mb_deriv_strat')
-                <- tcExtendTyVarEnv tvs $ do
+    do  { (mb_deriv_strat', deriv_tvs, (cls, cls_tys, cls_arg_kinds))
+                <- tcExtendTyVarEnv tvs $
                 -- Deriving preds may (now) mention
                 -- the type variables for the type constructor, hence tcExtendTyVarenv
                 -- The "deriv_pred" is a LHsType to take account of the fact that for
@@ -689,13 +685,8 @@ deriveTyData tvs tc tc_args mb_deriv_strat deriv_pred
                 -- Typeable is special, because Typeable :: forall k. k -> Constraint
                 -- so the argument kind 'k' is not decomposable by splitKindFunTys
                 -- as is the case for all other derivable type classes
-                     (deriv_tvs, cls, cls_tys, cls_arg_kinds)
-                       <- tcHsDeriv deriv_pred
-                     (strat_tvs, mb_deriv_strat')
-                       <- maybe ([], Nothing) (fmap Just) <$>
-                          traverse tcDerivStrategy mb_deriv_strat
-                     pure ( deriv_tvs, cls, cls_tys, cls_arg_kinds
-                          , strat_tvs, mb_deriv_strat' )
+                     tcDerivStrategy mb_deriv_strat $
+                     tcHsDeriv deriv_pred
 
         ; when (cls_arg_kinds `lengthIsNot` 1) $
             failWithTc (nonUnaryErr deriv_pred)
@@ -724,38 +715,54 @@ deriveTyData tvs tc tc_args mb_deriv_strat deriv_pred
         ; checkTc (enough_args && isJust mb_match)
                   (derivingKindErr tc cls cls_tys cls_arg_kind enough_args)
 
-        ; let Just kind_subst = mb_match
-              ki_subst_range  = getTCvSubstRangeFVs kind_subst
-              all_tkvs        = toposortTyVars $
-                                fvVarList $ unionFV
-                                  (tyCoFVsOfTypes tc_args_to_keep)
-                                  (FV.mkFVs deriv_tvs)
-              -- See Note [Unification of two kind variables in deriving]
-              unmapped_tkvs   = filter (\v -> v `notElemTCvSubst` kind_subst
-                                      && not (v `elemVarSet` ki_subst_range))
-                                       all_tkvs
-              (subst, _)      = mapAccumL substTyVarBndr
-                                          kind_subst unmapped_tkvs
-              final_tc_args   = substTys subst tc_args_to_keep
-              final_cls_tys   = substTys subst cls_tys
-              tkvs            = tyCoVarsOfTypesWellScoped $
-                                final_cls_tys ++ final_tc_args
+        ; let propagate_subst kind_subst tkvs' cls_tys' tc_args'
+                = (final_tkvs, final_cls_tys, final_tc_args)
+                where
+                  ki_subst_range  = getTCvSubstRangeFVs kind_subst
+                  -- See Note [Unification of two kind variables in deriving]
+                  unmapped_tkvs   = filter (\v -> v `notElemTCvSubst` kind_subst
+                                         && not (v `elemVarSet` ki_subst_range))
+                                           tkvs'
+                  (subst, _)      = mapAccumL substTyVarBndr
+                                              kind_subst unmapped_tkvs
+                  final_tc_args   = substTys subst tc_args'
+                  final_cls_tys   = substTys subst cls_tys'
+                  final_tkvs      = tyCoVarsOfTypesWellScoped $
+                                    final_cls_tys ++ final_tc_args
 
-        ; case mb_deriv_strat' of
-            Just (ViaStrategy via_ty) -> do
-              let via_kind = typeKind via_ty
-                  -- TODO RGS: It's not really "final", is it?
-                  final_inst_ty_kind = typeKind (mkTyConApp tc final_tc_args)
-                  via_match = tcUnifyTy final_inst_ty_kind via_kind
+        ; let tkvs = toposortTyVars $ fvVarList $
+                     unionFV (tyCoFVsOfTypes tc_args_to_keep)
+                             (FV.mkFVs deriv_tvs)
+              Just kind_subst = mb_match
+              (tkvs', final_cls_tys', final_tc_args')
+                = propagate_subst kind_subst tkvs cls_tys tc_args_to_keep
 
-              checkTc (isJust via_match)
-                      (derivingViaKindErr cls final_inst_ty_kind
-                                          via_ty via_kind)
-              -- TODO RGS: Actually apply this substitution
-            _ -> pure ()
+        ; (tkvs, final_cls_tys, final_tc_args, final_mb_deriv_strat) <-
+            case mb_deriv_strat' of
+              Just (ViaStrategy via_ty) -> do
+                let final_via_ty   = via_ty
+                    final_via_kind = typeKind final_via_ty
+                    final_inst_ty_kind
+                              = typeKind (mkTyConApp tc final_tc_args')
+                    via_match = tcUnifyTy final_inst_ty_kind final_via_kind
+
+                checkTc (isJust via_match)
+                        (derivingViaKindErr cls final_inst_ty_kind
+                                            final_via_ty final_via_kind)
+
+                let Just via_subst = via_match
+                    (final_tkvs, final_cls_tys, final_tc_args)
+                      = propagate_subst via_subst tkvs'
+                                        final_cls_tys' final_tc_args'
+                pure ( final_tkvs, final_cls_tys, final_tc_args
+                     , Just $ ViaStrategy $ substTy via_subst via_ty
+                     )
+
+              _ -> pure ( tkvs', final_cls_tys', final_tc_args'
+                        , mb_deriv_strat' )
 
         ; traceTc "Deriving strategy (deriving clause)" $
-            vcat [ppr mb_deriv_strat', ppr deriv_pred]
+            vcat [ppr final_mb_deriv_strat, ppr deriv_pred]
 
         ; traceTc "derivTyData1" (vcat [ pprTyVars tvs, ppr tc, ppr tc_args
                                        , ppr deriv_pred
@@ -785,7 +792,7 @@ deriveTyData tvs tc tc_args mb_deriv_strat deriv_pred
 
         ; spec <- mkEqnHelp Nothing tkvs
                             cls final_cls_tys tc final_tc_args
-                            Nothing mb_deriv_strat'
+                            Nothing final_mb_deriv_strat
         ; traceTc "derivTyData" (ppr spec)
         ; return $ Just spec } }
 

--- a/compiler/typecheck/TcHsType.hs
+++ b/compiler/typecheck/TcHsType.hs
@@ -6,6 +6,7 @@
 -}
 
 {-# LANGUAGE CPP, TupleSections, MultiWayIf, RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
 
 module TcHsType (
         -- Type signatures
@@ -244,7 +245,7 @@ tc_hs_sig_type (HsIB { hsib_vars = sig_vars
        ; return (mkSpecForAllTys tkvs ty) }
 
 -----------------
-tcHsDeriv :: LHsSigType GhcRn -> TcM ([TyVar], Class, [Type], [Kind])
+tcHsDeriv :: LHsSigType GhcRn -> TcM ([TyVar], (Class, [Type], [Kind]))
 -- Like tcHsSigType, but for the ...deriving( C t1 ty2 ) clause
 -- Returns the C, [ty1, ty2, and the kinds of C's remaining arguments
 -- E.g.    class C (a::*) (b::k->k)
@@ -261,20 +262,37 @@ tcHsDeriv hs_ty
        ; let (tvs, pred) = splitForAllTys ty
        ; let (args, _) = splitFunTys cls_kind
        ; case getClassPredTys_maybe pred of
-           Just (cls, tys) -> return (tvs, cls, tys, args)
+           Just (cls, tys) -> return (tvs, (cls, tys, args))
            Nothing -> failWithTc (text "Illegal deriving item" <+> quotes (ppr hs_ty)) }
 
-tcDerivStrategy :: DerivStrategy GhcRn
-                   -- TODO RGS: Explain what these type variables are
-                -> TcM ([TyVar], DerivStrategyPostTc)
-tcDerivStrategy StockStrategy    = pure ([], StockStrategy)
-tcDerivStrategy AnyclassStrategy = pure ([], AnyclassStrategy)
-tcDerivStrategy NewtypeStrategy  = pure ([], NewtypeStrategy)
-tcDerivStrategy (ViaStrategy ty) = do
-  cls_kind <- newMetaKindVar
-  ty' <- checkNoErrs $ tc_hs_sig_type_and_gen ty cls_kind
-  let (tvs, pred) = splitForAllTys ty'
-  pure (tvs, ViaStrategy pred)
+-- TODO RGS: Seriously, explain this.
+tcDerivStrategy :: forall a.
+                   Maybe (DerivStrategy GhcRn)
+                -> TcM ([TyVar], a)
+                -> TcM (Maybe DerivStrategyPostTc, [TyVar], a)
+tcDerivStrategy mds thing_inside
+  = case mds of
+      Nothing -> boring_case Nothing
+      Just ds -> do (ds', tvs, thing) <- tc_deriv_strategy ds
+                    pure (Just ds', tvs, thing)
+  where
+    tc_deriv_strategy :: DerivStrategy GhcRn
+                      -> TcM (DerivStrategyPostTc, [TyVar], a)
+    tc_deriv_strategy StockStrategy    = boring_case StockStrategy
+    tc_deriv_strategy AnyclassStrategy = boring_case AnyclassStrategy
+    tc_deriv_strategy NewtypeStrategy  = boring_case NewtypeStrategy
+    tc_deriv_strategy (ViaStrategy ty) = do
+      cls_kind <- newMetaKindVar
+      ty' <- checkNoErrs $ tc_hs_sig_type_and_gen ty cls_kind
+      let (via_tvs, via_pred) = splitForAllTys ty'
+      tcExtendTyVarEnv via_tvs $ do
+        (thing_tvs, thing) <- thing_inside
+        pure (ViaStrategy via_pred, via_tvs ++ thing_tvs, thing)
+
+    boring_case :: mds -> TcM (mds, [TyVar], a)
+    boring_case mds = do
+      (thing_tvs, thing) <- thing_inside
+      pure (mds, thing_tvs, thing)
 
 tcHsClsInstType :: UserTypeCtxt    -- InstDeclCtxt or SpecInstCtxt
                 -> LHsSigType GhcRn

--- a/testsuite/tests/deriving/should_compile/deriving-via-compile.hs
+++ b/testsuite/tests/deriving/should_compile/deriving-via-compile.hs
@@ -9,6 +9,7 @@
 {-# LANGUAGE GADTs #-}
 {-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
 {-# LANGUAGE MultiWayIf #-}
 {-# LANGUAGE TypeOperators #-}
 {-# LANGUAGE ScopedTypeVariables #-}
@@ -20,6 +21,7 @@ module DerivingViaCompile where
 
 import Data.Void
 import Data.Complex
+import Data.Functor.Const
 import Data.Functor.Identity
 import Data.Ratio
 import Control.Monad.Reader
@@ -28,7 +30,6 @@ import Control.Monad.Writer
 import Control.Applicative hiding (WrappedMonad(..))
 
 import Data.Bifunctor
-import Data.Functor.Identity
 import Data.Monoid
 import Data.Kind
 
@@ -121,7 +122,7 @@ instance MFunctor (StateT s :: MTrans) where
 
 instance MFunctor (WriterT w :: MTrans) where
   hoist :: Monad m => (m ~> m') -> (WriterT w m ~> WriterT w m')
-  hoist nat = WriterT . nat . runWriterT 
+  hoist nat = WriterT . nat . runWriterT
 
 infixr 9 `ComposeT`
 newtype ComposeT :: MTrans -> MTrans -> MTrans where
@@ -135,7 +136,7 @@ instance (MonadTrans f, MonadTrans g, LiftingMonad g) => MonadTrans (ComposeT f 
 
 instance (MFunctor f, MFunctor g, LiftingMonad g) => MFunctor (ComposeT f g) where
   hoist :: forall m m'. Monad m => (m ~> m') -> (ComposeT f g m ~> ComposeT f g m')
-  hoist f = ComposeT . hoist (hoist f) . getComposeT 
+  hoist f = ComposeT . hoist (hoist f) . getComposeT
     \\ proof @g @m
 
 -----
@@ -177,7 +178,7 @@ newtype P6 a = P6 [a]           deriving Show via ([] $ a)
 newtype P7 a = P7 (a, a)        deriving Show via (Identity $ (a, a))
 newtype P8 a = P8 (Either () a) deriving Functor via (($) (Either ()))
 
-newtype f $ a = APP (f a) deriving newtype Show deriving newtype Functor 
+newtype f $ a = APP (f a) deriving newtype Show deriving newtype Functor
 
 ----
 -- From Baldur's notes
@@ -245,7 +246,7 @@ data Sorted a = Sorted a a a
     via (WrapMonad Sorted)
   deriving (Num, Fractional, Floating, Semigroup, Monoid)
     via (WrapApplicative Sorted a)
-    
+
 
 instance Monad Sorted where
   (>>=) :: Sorted a -> (a -> Sorted b) -> Sorted b
@@ -257,7 +258,7 @@ instance Monad Sorted where
 instance Pointed Sorted where
   pointed :: a -> Sorted a
   pointed a = Sorted a a a
-  
+
 ----
 -- 3
 ----
@@ -271,7 +272,7 @@ newtype WrappedNumEq2 a = WrappedNumEq2 a
 instance (Num a, Eq a) => IsZero (WrappedNumEq a) where
   isZero :: WrappedNumEq a -> Bool
   isZero (WrappedNumEq a) = 0 == a
-  
+
 instance Show a => IsZero (WrappedShow a) where
   isZero :: WrappedShow a -> Bool
   isZero (WrappedShow a) = "0" == show a
@@ -280,7 +281,7 @@ instance (Num a, Eq a) => IsZero (WrappedNumEq2 a) where
   isZero :: WrappedNumEq2 a -> Bool
   isZero (WrappedNumEq2 a) = a + a == a
 
-newtype INT = INT Int 
+newtype INT = INT Int
   deriving newtype Show
   deriving IsZero via (WrappedNumEq Int)
 
@@ -292,8 +293,8 @@ newtype VOID = VOID Void deriving IsZero via (WrappedShow Void)
 class Bifunctor p => Biapplicative p where
   bipure :: a -> b -> p a b
 
-  biliftA2 
-    :: (a  -> b  -> c) 
+  biliftA2
+    :: (a  -> b  -> c)
     -> (a' -> b' -> c')
     -> p a a'
     -> p b b'
@@ -443,3 +444,17 @@ deriving via (WrapAdditive Maybe)    instance Affine Maybe
 deriving via (WrapAdditive ZipList)  instance Affine ZipList
 -- ADDITIVE(Identity)
 deriving via (WrapAdditive Identity) instance Affine Identity
+
+----
+-- 8
+----
+
+class C2 a b c where
+  c2 :: a -> b -> c
+
+instance C2 a b (Const a b) where
+  c2 x _ = Const x
+
+newtype Fweemp a = Fweemp a
+  deriving (C2 a b)
+       via (Const a (b :: Type))

--- a/testsuite/tests/deriving/should_fail/all.T
+++ b/testsuite/tests/deriving/should_fail/all.T
@@ -69,3 +69,4 @@ test('T12512', omit_ways(['ghci']), compile_fail, [''])
 test('T12801', normal, compile_fail, [''])
 test('T14365', [extra_files(['T14365B.hs','T14365B.hs-boot'])],
                multimod_compile_fail, ['T14365A',''])
+test('deriving-via-fail', normal, compile_fail, [''])

--- a/testsuite/tests/deriving/should_fail/deriving-via-fail.hs
+++ b/testsuite/tests/deriving/should_fail/deriving-via-fail.hs
@@ -1,0 +1,11 @@
+{-# LANGUAGE DerivingVia #-}
+module DerivingViaFail where
+
+import Control.Category
+import Data.Functor.Identity
+
+newtype Foo1 a = Foo1 a deriving Show via (Identity b)
+
+newtype Foo2 a b = Foo2 (a -> b)
+  deriving Category
+    via fooo

